File: server/utils/weather_fetcher.py
import requests
from datetime import datetime
import json
from typing import Dict, Any, Optional
import os
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

class WeatherFetcher:
    """Weather data fetcher using the WeatherAPI"""
    
    def __init__(self, api_key: str):
        """Initialize with API key"""
        self.api_key = api_key
        self.base_url = "http://api.weatherapi.com/v1"
    
    def _make_request(self, endpoint: str, params: Dict[str, Any]) -> Dict[str, Any]:
        """Make a request to the WeatherAPI"""
        # Add API key to params
        params['key'] = self.api_key
        
        # Make request
        response = requests.get(f"{self.base_url}/{endpoint}", params=params)
        
        # Check for errors
        response.raise_for_status()
        
        # Return JSON response
        return response.json()
    
    def fetch_weather(self, city: str, lang: str = 'en') -> Dict[str, Any]:
        """Fetch current weather for a city"""
        params = {
            'q': city,
            'lang': lang,
            'aqi': 'no'  # Don't include air quality data
        }
        return self._make_request('current.json', params)
    
    def fetch_forecast(self, city: str, lang: str = 'en', days: int = 3) -> Dict[str, Any]:
        """Fetch weather forecast for a city"""
        params = {
            'q': city,
            'lang': lang,
            'days': days,
            'aqi': 'no'
        }
        return self._make_request('forecast.json', params)
    
    def fetch_alerts(self, city: str, lang: str = 'en') -> Dict[str, Any]:
        """Fetch weather alerts for a city"""
        params = {
            'q': city,
            'lang': lang,
            'alerts': 'yes'
        }
        return self._make_request('forecast.json', params) 

File: server/utils/weather_fetcher.ts
import type { WeatherData, WeatherForecast } from '@/types/weather'

export class WeatherFetcher {
  private apiKey: string
  private baseUrl = 'https://api.weatherapi.com/v1'

  constructor(apiKey: string) {
    if (!apiKey) {
      throw new Error('Weather API key is required')
    }
    this.apiKey = apiKey
  }

  private async fetchWithErrorHandling(url: string): Promise<any> {
    try {
      const response = await fetch(url, {
        headers: {
          'Accept': 'application/json'
        }
      })
      
      if (!response.ok) {
        const contentType = response.headers.get('content-type')
        if (contentType?.includes('application/json')) {
          const errorData = await response.json()
          throw new Error(errorData.error?.message || `Weather API error: ${response.statusText}`)
        } else {
          throw new Error(`Weather API error: ${response.status} ${response.statusText}`)
        }
      }

      const contentType = response.headers.get('content-type')
      if (!contentType?.includes('application/json')) {
        throw new Error('Invalid response format: Expected JSON')
      }

      return response.json()
    } catch (error) {
      if (error instanceof Error) {
        if (error.message.includes('API key')) {
          throw new Error('Invalid API key')
        }
        if (error.message.includes('No matching location')) {
          throw new Error('City not found')
        }
        throw error
      }
      throw new Error('Failed to fetch weather data')
    }
  }

  async getCurrentWeather(city: string): Promise<WeatherData> {
    if (!city) {
      throw new Error('City parameter is required')
    }

    const url = `${this.baseUrl}/current.json?key=${this.apiKey}&q=${encodeURIComponent(city)}&aqi=yes&alerts=yes`
    return this.fetchWithErrorHandling(url)
  }

  async getForecast(city: string, days: number): Promise<WeatherForecast> {
    if (!city) {
      throw new Error('City parameter is required')
    }
    if (days < 1 || days > 3) {
      throw new Error('Days parameter must be between 1 and 3')
    }

    const url = `${this.baseUrl}/forecast.json?key=${this.apiKey}&q=${encodeURIComponent(city)}&days=${days}&aqi=yes&alerts=yes`
    const data = await this.fetchWithErrorHandling(url)
    
    // Log the forecast data for debugging
    console.log('Forecast data received:', {
      city,
      days,
      forecastDays: data.forecast?.forecastday?.length || 0,
      firstDay: data.forecast?.forecastday?.[0]?.date,
      lastDay: data.forecast?.forecastday?.[data.forecast?.forecastday?.length - 1]?.date
    })

    return data
  }
} 